### PR TITLE
[Enhancement] Added possibility to specify a batch size to the Particle Container

### PR DESCRIPTION
**Pull request by bchevalier**
It Improves performance as the user can now specify a batch size for the particles. It is especially useful when developing for slow hardware such as mobile devices.

Note: I plan on doing another optimisation on top of it that should (crossing fingers) allow developers to greatly improve performance when it comes to adding and removing particles from the container (it should show good results on the bunnymark).

### CODE REVIEW

File: src/core/particles/ParticleContainer.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ var Container = require('../display/Container'),
  * @extends PIXI.Container
  * @memberof PIXI
  *
- * @param [size=15000] {number} The number of images in the SpriteBatch before it flushes.
+ * @param [totalSize=15000] {number} The maximum number of particles that can be renderer by the container.
```
Comment (englercj) [inline src/core/particles/ParticleContainer.js]: `size` or `maxSize` is better naming for this I think.

File: src/core/particles/ParticleContainer.js [hunk 2]
```diff
@@ -33,10 +33,25 @@ var Container = require('../display/Container'),
  * @param [properties.uvs=false] {boolean} When true, uvs be uploaded and applied.
  * @param [properties.alpha=false] {boolean} When true, alpha be uploaded and applied.
  */
```
Comment (englercj) [inline src/core/particles/ParticleContainer.js]: Need to add a line for `batchSize`.

**Comment by englercj** [comment]
Good start, this adds the code necessary for initializing the buffers to a different size if you want. It doesn't add the dynamic resizing that we talked about yet, but that is OK we can add it later.

File: src/core/particles/ParticleContainer.js [hunk 3]
```diff
@@ -33,10 +33,25 @@ var Container = require('../display/Container'),
...
-function ParticleContainer(size, properties)
+function ParticleContainer(totalSize, properties, batchSize)
 {
     Container.call(this);
 
+    batchSize = batchSize || 15000; //CONST.SPRITE_BATCH_SIZE; // 2000 is a nice balance between mobile / desktop
```
Comment (englercj) [inline src/core/particles/ParticleContainer.js]: Why did the default change to 15k? The problem this is trying to solve is not updating a large buffer that is mostly unused, but this actually will make that worse for most people by default.

File: src/core/particles/ParticleContainer.js (hunk 2 above)
Comment (bchevalier) [inline src/core/particles/ParticleContainer.js]: yes, good catch

File: src/core/particles/ParticleContainer.js (hunk 1 above)
Comment (bchevalier) [inline src/core/particles/ParticleContainer.js]: I will change to maxSize

File: src/core/particles/ParticleContainer.js (hunk 3 above)
Comment (bchevalier) [inline src/core/particles/ParticleContainer.js]: It was already 15k but it used to be in the ParticleRenderer component (see ```this.size = 15000```).

**Comment by bchevalier** [comment]
I realised that there would be quite some changes to do in order to make the buffer size grow. This solution was more elegant and it fits my needs so I guess it fits the needs of people with similar performance requirements. I think this can substitute to a dynamically resizing buffer.

**Comment by englercj** [comment]
Yeah like I said, I think this is a great first step in that direction. This alone is very helpful.

File: src/core/particles/ParticleContainer.js (hunk 3 above)
Comment (englercj) [inline src/core/particles/ParticleContainer.js]: Oh my bad, you are right!

**Comment by englercj** [comment]
Looks good to me, would you mind squashing these commits and I'll merge it in?

```
git rebase -i HEAD~6
git push -f origin particleContainerRefactor3
```

Should do the trick!

**Comment by bchevalier** [comment]
After the rebase it tells me that everything is already up to date so I cannot push anything

**Comment by englercj** [comment]
It should say that your branches have diverged and make you need to force push.

Thats fine, I can rebase it and merge it manually.

**Comment by englercj** [comment]
Manually rebased and merged.

File: src/core/particles/webgl/ParticleRenderer.js [hunk 4]
```diff
@@ -270,9 +269,9 @@ ParticleRenderer.prototype.generateBuffers = function ( container )
         this.properties[i].dynamic = container._properties[i];
     }
 
-    for (i = 0; i < size; i += this.size)
+    for (i = 0; i < size; i += batchSize)
     {
-        buffers.push( new ParticleBuffer(gl,  this.properties, this.size, this.shader) );
+        buffers.push( new ParticleBuffer(gl,  this.properties, batchSize, this.shader) );
```
Comment (bchevalier) [inline src/core/particles/webgl/ParticleRenderer.js]: I believe a small mistake was made when resolving the conflict here, if you look at the current dev version ```this.size``` hasn't been replaced by ```batchSize```

**Comment by englercj** [comment]
Good catch, thanks for that! Let me know if you see anything else.

**Pull request closed without merging**